docs/ref/contrib/postgres/indexes.txt

```diff
@@ -78,3 +78,24 @@ available from the ``django.contrib.postgres.indexes`` module.
 
     .. _buffering build: https://www.postgresql.org/docs/current/static/gist-implementation.html#GIST-BUFFERING-BUILD
     .. _fillfactor: https://www.postgresql.org/docs/current/static/sql-createindex.html#SQL-CREATEINDEX-STORAGE-PARAMETERS
+
+``HashIndex``
+=============
+
+.. class:: HashIndex(fillfactor=None, **options)
+
+    .. versionadded:: 2.1
+
+    Creates a `hash index
+    <https://www.postgresql.org/docs/current/static/hash.html>`_.
+
+    Provide an integer value from 10 to 100 to the fillfactor_ parameter to
+    tune how packed the index pages will be. PostgreSQL's default is 90.
+
+    .. note:
+
+        Although hash indexes have been available in PostgreSQL for a long
```

Comment:
I'm unsure about this. As far as I see, the tests pass on older versions.

Reply:
Yes, they will. Hash indexes work on older versions of PostgreSQL, but they were unsafe, and I thought it would be irresponsible to allow people to use them unless they are using version 10 or later. If we decide that Django shouldn't police this, then we still need a big fat warning.

See the first caution at https://www.postgresql.org/docs/9.6/static/indexes-types.html

Also read this article: http://rhaas.blogspot.co.uk/2017/09/postgresqls-hash-indexes-are-now-cool.html

Reply:
Oh, I didn't notice the subsequent commit in this PR that added the version check -- I would have put that in the same commit. I don't know if someone might want to use hash indexes on older versions of PostgreSQL but it doesn't really feel like Django's place to block usage of it.

Reply:
Sorry - the version checks I moved into subsequent commits so that they could be dropped if decided against. This was based on previous discussion with you above - I'm not sure we've had a firm decision on that yet since my most recent comments that the errors returned are too vague when index types or options do not exist.

We could allow hash indexes with <10 but they don't replicate, or backup properly IIRC, and if PostgreSQL crashes they are highly likely to become corrupt. This makes their use quite limited...

Reply:
For Hash indexes, I think a documentation warning is sufficient. It seems unlikely that someone would choose an index without reading about how it works. I guess the other version checks are fine.

Reply:
Ok. We'll go with that then.